### DISCUSSION TITLE
How to implement website testing in parallel with GitHub Actions?

### DISCUSSION THREAD

**Discussion by ptmkenny** [community/community#25239] [Actions]
<p>I have a Drupal 9 site that I have set up a GitHub Actions workflow for in order to run some tests:</p>
<ul>
<li>pa11y</li>
<li>behat</li>
<li>playwright</li>
</ul>
<p>The problem is that it takes about 5 hours to run all the tests because they are executed one-by-one.  So now I want to execute them in parallel, but I want to avoid duplicating my setup configuration.</p>
<p>It appears that GHA supports running <strong>jobs</strong> in parallel, but not <strong>steps</strong>.  (Please correct me if I’m wrong.)</p>
<p>So, I can break my tests up into several jobs:</p>
<ul>
<li>pa11y</li>
<li>behat part 1</li>
<li>behat part 2</li>
<li>behat part 3</li>
<li>playwright part 1</li>
<li>playwright part 2</li>
</ul>
<p>Here’s where I have an issue: My workflow has about 20 steps that set up the test environment (run composer install, run npm install, generate some variables for the tests, and so on) for running all of the tests.  If possible, I don’t want to copy this code for each job, because I know that later I will probably make a mistake changing it in some places and not others.</p>
<p>So, how can I reuse the steps to set up the test environment for each job?  I read about <a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows" rel="noopener nofollow ugc">reusable workflows</a>, but is this the way to go?  (Create a reusable workflow to set up the test environment, and then call that workflow for each part of the tests?)</p>
<p>If so, is there a way to combine all the artifacts from different jobs into a single zipped archive?</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="ptmkenny" data-post="1" data-topic="239899">
<div class="title">
<div class="quote-controls"></div>
 ptmkenny:</div>
<blockquote>
<p>My workflow has about 20 steps that set up the test environment (run composer install, run npm install, generate some variables for the tests, and so on) for running all of the tests. If possible, I don’t want to copy this code for each job, because I know that later I will probably make a mistake changing it in some places and not others.</p>
</blockquote>
</aside>
<p>That sounds like a <a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-action">composite action</a> would make sense here, it’s something I’ve done for one of my repositories, too (in that case as a local action stored within the repository). Reference the action in each of the jobs, and the steps will be called in its position.</p>
<aside class="quote no-group" data-username="ptmkenny" data-post="1" data-topic="239899">
<div class="title">
<div class="quote-controls"></div>
 ptmkenny:</div>
<blockquote>
<p>If so, is there a way to combine all the artifacts from different jobs into a single zipped archive?</p>
</blockquote>
</aside>
<p>Well, if you make another job that downloads and combines them. Up to you whether it’s worth the effort. If you do you might want to set a short retention time for the partial artifacts.</p>

**Comment by ptmkenny** [comment]
<p>Thanks, I successfully implemented this as a composite action.  This is way easier and cleaner than before!</p>
